docs/design/eth-subscribe.md

```diff
@@ -0,0 +1,335 @@
+|:-------------:|:------------------------------------------------------------------------------------------------------------------------:|:---------------------------------------------------------------------------------------------------------------------------------------:|
+|     32607     | The maximum batch size that can be sent over a websocket connection is 10. Please decrease the batch size and try again. |                             Occurs when user attempts to send high-volume JSON-RPC traffic over Websockets.                             |
+|     32608     |                       The maximum number of socket connections is reached. Please try again later.                       |     Occurs when a new connection is established, but that exceeds the limit of allowed live connections (controlled by an env var).     |
+|     32609     |                                                 Invalid subscription id                                                  | Occurs when user tries to unsubscribe from a non-existing subscription or one that is not associated with the active socket connection. |
+|     32610     |                                    Invalid subscription parameters: {detailed reason}                                    |               Occurs when there is an error due to the provided subscription parameters, i.e. the address does not exist.               |
+|     32611     |                                                   Subscription aborted                                                   |                          Occurs when the subscription is aborted by the relay, i.e. a maximum TTL is reached.                           |
+
+## Non-Functional Requirements
+
+WebSockets in general are resource intensive. We could implement a throttling mechanism. Every live socket connection would also result in additional requests to the Mirror Node every n seconds. This means that we need to introduce a rate limiting feature, like 5 concurrent connections by IP.
+
+1. Implement an IP based limit for maximum concurrent connections.
+2. Implement a global limit for maximum concurrent connections.
+3. Explore the possibility of adding a TTL limit for every connection.
+4. Explore the option to add a limit for the total subscribed time per IP ( `sum(connection * subscribed_time)` ). There could theoretically be a way to charge users to increase this limit, for example through via a smart contract. Alternatively users could be encouraged to run their own instances of the relay.
+5. Users should be required to renew their subscription in the case of an error. If for some reason the relay is restarted it should not automatically try to restore all previous connections and subscriptions.
+
```

Comment:
Should we also explore the possibility of auto-ending some subscriptions when they are responded, for example:
 1. User subsribes to a transaction by hash, that is on pending state.
 2. Relay receives the transaction once is mined and included in a block and returns the value to the subscriber.
 3. since that tx is final, is certain there wont be more changes, user should unsubscribe, but in case he does not, we automatically do at the time of response.

Reply:
I already answered this from the meeting.